### ISSUE TITLE
Proposal: Allow dynamic dependency override

### ISSUE THREAD

**Issue by vdice** [helm/helm#2205]
**Proposal:** Enable overriding one or more 'umbrella' chart dependencies with specific sub-chart version(s) and/or sub-chart repo(s) at chart installation time.

Updated `install` command might look something like:
```
$ helm install umbrella-chart-repo/umbrella-chart \
   --set-requirement sub-chart-name.version=specific-chart-version \
   --set-requirement sub-chart-name.repository=specific-chart-repo
```
(If `sub-chart-name.repository` not provided, would use value currently existing in umbrella chart's `requirements.yaml`)

Alternatively, perhaps one could supply an entirely new `requirements.yaml` file and supply it at install time in a manner similar to `-f values.yaml`:
```
$ helm install umbrella-chart-repo/umbrella-chart \
   -r custom-requirements.yaml
```
To support this latter case, it would most likely warrant a separate ticket adding a way to inspect an umbrella chart's requirements à la `helm inspect values umbrella-chart-repo/umbrella-chart`:
```
$ helm inspect requirements umbrella-chart-repo/umbrella-chart > default-requirements.yaml

$ # modify file as necessary and save as custom-requirements.yaml

$ helm install umbrella-chart-repo/umbrella-chart \
   -r custom-requirements.yaml
```

**Considerations:** 

The details of supporting this may be tricky.  At a high level, I would imagine `helm install` would trigger a `helm dep update` behind the scenes, perhaps staging this updated (and packaged?) chart locally and then installing.  The bulk of the considerations/implications around this proposal may indeed lie in the details therein.

I would assume `helm upgrade` would be updated to mirror the functionality added to `helm install`.

**Similar Issues**:
 * https://github.com/kubernetes/helm/issues/1947 requests the addition of a `helm dependency create/add ...` command but it seems to be centered around chart build/package/publish time and so doesn't satisfy the dynamic needs of this request.
 * https://github.com/kubernetes/helm/issues/2183 requests a "development" mode wherein invocations of `helm install` would use locally staged/updated charts dynamically.  This is closer in spirit but with the fundamental difference that this proposal would enable fetching _remote_ charts as opposed to ones existing locally.  (Support for both would be ideal.)

**Use Cases**:

I'll mention the use case that triggered this feature request.  I work on CI in support of publishing and releasing an umbrella chart dependent on many sub-charts.  

For testing a PR involving a change to a given sub-chart, we first publish the sub-chart artifact and we then publish a correlating umbrella chart artifact after injecting the said sub-chart version and repo into the `requirements.yaml` and running `helm dependency update`.  Lastly, we run our E2E tests against the umbrella chart artifact.

This pipeline could be simplified by removing the need to publish a new umbrella chart artifact for every change in a given sub-chart.  We'd only need to publish the sub-chart artifact and then run our E2E tests by installing the default umbrella chart but with the updated `requirements.yaml` specified at runtime.

This use case could be extrapolated into a general enhancement/UX upgrade for dev/hacking on a given umbrella chart's dependencies.  It also supports the ability to quickly substitute one or more sub-chart versions for trouble-shooting/debugging/etc.

**Comment by technosophos** [comment]
I would think that passing a custom `requirements.yaml` would be relatively easy to do. And once that is done, adding a `--set`-like system like you propose also seems doable.

**Comment by gmorse81** [comment]
/reopen
/remove-lifecycle rotten

I am very interested in this as a helm user. i would offer to contribute, but would need some assistance on a starting point.

**Comment by FX-HAO** [comment]
I implemented this feature. It works on my mac, but I think it needs more tests before making a PR. Can someone help me to test it and give me some feedback? repo: https://github.com/FX-HAO/helm

Usage:
```
bin/helm install -r custom-requirements.yaml chart-name 
bin/helm upgrade -r custom-requirements.yaml chart-name
```
Or you can specify `none` to remove all requirements.

**Comment by vdice** [comment]
@FX-HAO Nice work!  I'll see if I can get some cycles to test this from your fork, though I'd recommend creating a PR if you are getting close to feeling ready.

To note, if you hadn't seen already, it looks like Helm v3 will have native support for the high-level functionality mentioned here via the [extension scripts](https://github.com/helm/community/blob/master/helm-v3/001-charts.md#the-ext-directory) feature.

**Comment by xbmono** [comment]
I look forward to see this. We need this for our continous delivery process

**Comment by djuarezg** [comment]
We are really interested in this feature.

**Comment by vcorr** [comment]
Apparently Helm 3 does not provide a solution for this(?). I would really like to see a solution for this

**Comment by bacongobbler** [comment]
I'm going to go ahead and mark this as a "Help Wanted". Please feel free to submit your ideas to implement this functionality. It's been over 3 years since this proposal was first opened and some things have changed since then (requirements.yaml was deprecated, for example), so it might be worth taking a closer look at the internals first before submitting a PR.

**Comment by papanito** [comment]
This would be really helpful, especially for environments where you cannot download from external repos but only from internal "proxy" repos. Not sure if I can contribute anything as I don't know go

**Comment by bacongobbler** [comment]
see #7312

**Comment by matthiaslee** [comment]
We would be very interested in this feature and would avoid us needing to copy a chart just to update a dependency version.

**Comment by mladedav** [comment]
I started implementing this in #9540. I think it works in install but I would like some feedback from maintainers before I put it in upgrade also.

**Comment by mladedav** [comment]
I found a few issues that should probably be clarified now.

If someone specifies the `--set-dependency` option, should `--dependency-update` option be automatically also included? I would guess so, but according to the help text, it should call `helm dependency update` in the background, which would update the contents of the `charts` directory with dependencies, but it would put the changed versions there, it would probably update the lock file (or we could bypass that one), while the `Chart.yaml` would still list the older dependencies. Subsequent install/update commands do not check the versions, only if the dependency has the same name, therefore it would keep using the older dependency, until `helm dependency update` is called (directly or indirectly).
Another option is to use some non-standard directory for the dependencies for this run only (I have not checked if that is feasible) or avoid the changes in some other way.

Some of the questions boil down directly if this option is to be used for one-time changes only (that's the feeling I got from the issue) or if the changes should be persisted.

The questions for this proposal therefore are:
What should happen if `--set-dependency` needs to download a dependency, but `dependency-update` is not specified?
What should happen, if `--set-dependency` should install a new dependency, but no version is present?
What should happen, if `--set-dependency` should change a version of dependency, but another version is already present?
Should the downloaded dependency charts be persisted?
Should the Chart.yaml file be updated based on the command?
Should the Chart.lock file be updated based on the command?
Should the checks during install/upgrade also check versions and not only names of dependencies (I assume not as that would be a breaking change) and in that case what should happen if another version than the specified in `--set-dependency` is already present?

**Comment by technosophos** [comment]
This feature request should definitely be written up as a HIP. The HIP process has been online since late last year, and is much better for identifying potential issues. Thanks for the thoughtful response @mladedav .

On that note, I don't think @vdice is working on this anymore, so if someone else wants to start the HIP process, head on over here: https://github.com/helm/community/blob/master/hips/hip-0001.md

**Comment by furuholm** [comment]
Any progress on this? I can see many scenarios where this would be very valuable. I am particularly interested in being able to have a shared umbrella chart for an org, and then let individual teams override the chart they are working on until they release a version that is then includes in the shared umbrella chart.

**Comment by bacongobbler** [comment]
@mladedav started work on a HIP. You can find the link just above your comment (helm/community#176).

I would suggest reviewing that HIP and ask questions. It's good to have many community members review HIPs so we can be sure that a proposal can address many use cases.

**Comment by furuholm** [comment]
Thanks for the quick reply! My question was actually about the progress of the HIP. Sorry for not making that clear. However I gather from your response that you are waiting on community feedback. 

My initial reading of the HIP was 👍 but I will have a look at the HIP and see if there is anything I can contribute. There are a number of things that the HIP mentions needs to be sorted out, but I’m not knowledgeable enough to provide much of value there.

**Comment by mladedav** [comment]
@furuholm  As @bacongobbler said, I have written up the HIP and I would appreciate any feedback from anyone.

For the record I think that this feature will be written only for v4 (or so I've written in the HIP) because there are some issues concerning `--dependency-update` and this flag will be replaced in the next major version. Therefore the functionality still won't be available for some time.

**Comment by MitchAman** [comment]
> @furuholm As @bacongobbler said, I have written up the HIP and I would appreciate any feedback from anyone.

Thanks for writing up the HIP, @mladedav, I'd love to see this functionality come to fruition.  Have you started a discussion for this on [cncf-helm](https://lists.cncf.io/g/cncf-helm)?

**Comment by mladedav** [comment]
@MitchAman I have not, as I read HIP 0001 gave me the feeling that that is something to do to decide whether the HIP would be actually wanted which I decided was settled by the activity of this issue and I am not really used to email lists. I have sent an email just now asking for feedback on the HIP.

**Comment by loganknecht** [comment]
Hello.

Has there been any update on this?

My organization needs this feature for the same reasons as others above. We have an umbrella chart, but need to target specific dependency versions in order to perform testing in CI/CD. Specifically we are building a new candidate chart for the dependency and want to be able to use the umbrella chart to target that new candidate so that we can test it ahead of time.

Any update on this or potential work arounds would greatly be appreciated.

**Comment by gmorse81** [comment]
@loganknecht I created a pre-processor which does what you are looking for. Its probably overkill but it works. my team uses it for every helm deployment that we do. 

https://github.com/gmorse81/boondoggle

**Comment by eduardobaitello** [comment]
_Hello, I'm a further activity_

**Comment by eduardobaitello** [comment]
_Further activity to keep this alive...._

**Comment by LoicAG** [comment]
I would also be very interested in such a feature

**Comment by loganknecht** [comment]
This request is still desired. Please do not close this.

**Comment by vherzog-sauce** [comment]
Our organization would greatly benefit from this feature.

**Comment by papanito** [comment]
This request is still desired. Please do not close this.

**Comment by loganknecht** [comment]
This request is still desired. Please do not close this.

**Comment by ErezArbell** [comment]
I also need this feature.

**Comment by tobbyioa** [comment]
This feature is desired. Please, do not close this thread.

**Comment by JuryA** [comment]
Working on an amazingly simple and smart solution that implements the ability to merge/JSONpatch/... itself via `values.yaml`. The main component will be the Helm plugin which allows - simply said - "patch" whatever existing Helm chart "on the fly" so I can try to implement such a feature... I plan to release this project including sources as Open Source soon - just need to make some tests...

**Comment by kaktusztea** [comment]
Please don't close this.

**Comment by erik7z** [comment]
still required...

**Comment by loganknecht** [comment]
No not stale.

People still want this.

**Comment by ErezArbell** [comment]
This feature is needed.

**Comment by loganknecht** [comment]
Please do not close this.
This feature is needed.

**Comment by dhirschfeld** [comment]
Not stale.

**Comment by asarenkansah** [comment]
We need this! Please!!

**Comment by dgroh** [comment]
Not stale!

**Comment by PykupeJIbc** [comment]
We need this! Please!!!

**Comment by MaksimRudakov** [comment]
We need this! Please!!!

**Comment by dhirschfeld** [comment]
https://nostalebots.xyz

**Comment by oatnog** [comment]
I just want to update Chart.yaml's `appVersion` to match, for example, a docker tag I'm using to upgrade a helm-installed application. Maybe there's another way?

**Comment by dhirschfeld** [comment]
https://jacobtomlinson.dev/posts/2024/most-stale-bots-are-anti-user-and-anti-contributor-but-they-dont-have-to-be/

**Comment by loganknecht** [comment]
8 year old feature request, but yeah sure, it's stale 🙄

**Comment by loganknecht** [comment]
This is not stale, and still desired by the community.

Once again we are here again telling you not to close this.

![Image](https://github.com/user-attachments/assets/6ad7975d-00c1-456b-b944-5f1cfa67ca61)

**Comment by dhirschfeld** [comment]
The ball is clearly in the maintainers court to implement this much wanted feature. Spamming everyone every few months simply because no maintainer has had time to implement it yet is anti-user.

Please turn off your spam bot, or configure it to only spam where the issue is missing some information necessary for the maintainers to action it - i.e. only when the ball is in the users court to update the issue, or otherwise take some meaningful action, other than just repeated `ACK`'s every few months.

https://jacobtomlinson.dev/posts/2024/most-stale-bots-are-anti-user-and-anti-contributor-but-they-dont-have-to-be/

**Comment by TerryHowe** [comment]
Projects like this are a community effort and it isn't maintainers responsibility to implement every/any change.  Anyone could raise a PR.

**Comment by dhirschfeld** [comment]
> Projects like this are a community effort and it isn't maintainers responsibility to implement every/any change. Anyone could raise a PR.

That is correct. I'm not asking anyone to implement anything. I'm just asking the maintainers to not inflict meaningless spam on the users who would like this feature implmented. It's as bad as a `+1` comment.

If there has been no change to the status of the issue, no need to say anything. That way, when we receive a notification from this issue it will be meaningful, and not just noise.

i.e. either configure the spam bot correctly, or just get rid of it sltogether - I doubt it's adding any actual value.

**Comment by loganknecht** [comment]
Not stale guys. C'mon now.

**Comment by ruslan-shaydullin** [comment]
Hi, I’d like to champion a new, narrowly scoped Helm 4 HIP for this feature. I’ve reviewed the closed PR #9540 and HIP #176, and I propose limiting the first iteration to overriding an existing direct dependency with an exact version from its already declared repository.

The override would apply to a single invocation without modifying `Chart.yaml`, `Chart.lock`, or `charts/`, with consistent behavior across `install`, `upgrade`, and `template`. Adding or removing dependencies, overriding repositories, and handling transitive dependencies would remain out of scope for now. Before any implementation begins, I’d like the HIP to define the semantics for provenance, `--verify`, the effective lock, and interaction with `--dependency-update`.

Is there still maintainer interest in pursuing this feature, and would a current maintainer be willing to shepherd and review the HIP? I’d also appreciate guidance on how this work should be ordered relative to #32588, HIP #441 / PR #32243, and HIP #443.
